tests/testthat/test-property.R

```diff
@@ -324,3 +324,29 @@ test_that("can validate with custom validator", {
     foo(x = 1:2)
   })
 })
+
+test_that("prop<- won't infinitly recurse on a custom setter", {
+  chattily_sync_ab <- function(self, value) {
+    cat("Starting syncup with value:", value, "\n")
+    a_value <- paste0("a_", value)
+    b_value <- paste0("b_", value)
+
+    cat(sprintf('setting @a <- "%s"\n', a_value))
+    self@a <- a_value
+
+    cat(sprintf('setting @b <- "%s"\n', b_value))
+    self@b <- b_value
```

Comment:
I'd like to revisit this test after the next OOP-WG meeting, to get confirmation that the current behavior is what we want before expanding the tests. 

Some questions for the next meeting:
- Should a property setter invoke other property setters from `self@foo<-` calls, or should class "internal" functions like a property setter default to taking the fast path?
- Should a constructor invoke custom property setters when constructing an instance?
- Should a constructor set default properties on an instance (properties with custom setters or otherwise), or should `prop()` resolve a default value them from the object parent class as needed?

Reply:
OOP-WG 2024-02-12 meeting discussion:

- Agreement that a custom property setter should not be called during instance initialization.

- Agreement that we should fetch a default property value from the parent class instead of requiring the default property attribute be attached to each instance, to allow for more light-weight instances.

- Agreement that "properties are attributes" is a stable implementation detail, and that `attr<-` is the escape-hatch for advanced users that "know what they're doing" that want to set a property value without invoking the property setter.

Reply:
I think this PR is good to merge (pending review from @DavisVaughan). 
We'll want the following in another PR:

- update `@` to be able to fetch default property value from the parent, 
- update the class constructor so that:
  - it doesn't invoking custom property setters on object instantiation.
  - doesn't attach default property values to instances. 
  - (maybe address some of the issues raised in https://github.com/RConsortium/S7/pull/341 while we're touching that part of the code, w.r.t. the class instance constructors signature)
- update/augment some of snapshot tests, so they're not _just_ snapshot tests (e.g., call `expect_equal()` or similar), to catch future regressions on `validate()`

Reply:
Sounds good!